File: 1.Arrays&Strings/1.2.cpp
// Check Permutation: Given two strings, write a method to decide if one is a permutation of the
// other.

#include "all.hpp"

bool checkPermutation(std::string &&one, std::string &&two)
{
    if (one.length() != two.length())
        return false;

    // I would use std::unordered_map instead of the array in case of std::wstring
    int charCount[256] = { 0 };

    for (int c : one)
        ++charCount[c];

    for (int c : two)
        if (--charCount[c] < 0)
            return false;

    return true;
}

bool checkPermutationStd(std::string &&one, std::string &&two)
{
    return std::is_permutation(one.begin(), one.end(), two.begin(), two.end());
}

void printResult(bool result, std::string &&one, std::string &&two)
{
    if (result)
        std::cout << '"' << two << '"' << " is permutation of " << '"' << one << '"'  << std::endl;
    else
        std::cout << '"' << two << '"' << " is not permutation of " << '"' << one << '"'  << std::endl;
}

int main()
{
    test(checkPermutation, printResult, "TestPermutation", "PermutationTest");
    test(checkPermutationStd, printResult, "TestPermutation", "PermutationTest");
    test(checkPermutation, printResult, "TestPermutation", "PeermutatTstion");
    test(checkPermutationStd, printResult, "TestPermutation", "PeermutatTstion");
    test(checkPermutation, printResult, "TestPermutation", "RotationTest");
    test(checkPermutationStd, printResult, "TestPermutation", "RotationTest");
    test(checkPermutation, printResult, "TestPermutation", "Permutationtest");
    test(checkPermutationStd, printResult, "TestPermutation", "Permutationtest");
    return 0;
}

File: 1.Arrays&Strings/1.2.rs
// Check Permutation: Given two strings, write a method to decide if one is a
// permutation of the other.

use googletest::prelude::*;

fn check_permutation(one: &str, two: &str) -> bool {
  if one.len() != two.len() {
    return false;
  }

  let mut char_count: [i32; 256] = [0; 256];
  for c in one.as_bytes() {
    char_count[*c as usize] += 1;
  }
  for c in two.as_bytes() {
    char_count[*c as usize] -= 1;
    if char_count[*c as usize] < 0 {
      return false;
    }
  }
  return true;
}

#[gtest]
fn check_permutation_assertions() {
  expect_true!(check_permutation("TestPermutation", "PermutationTest"));
  expect_true!(check_permutation("TestPermutation", "PeermutatTstion"));
  expect_false!(check_permutation("TestPermutation", "RotationTest"));
  expect_false!(check_permutation("TestPermutation", "Permutationtest"));
}
